staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -667,8 +669,35 @@ func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	c.startDispatching(event)
 
 	// Since add() can block, we explicitly add when cacher is unlocked.
+
+	// Dispatching event in nonblocking way first, which make faster watchers
+	// not be blocked by slower ones.
+	c.blockedWatchers = c.blockedWatchers[:0]
 	for _, watcher := range c.watchersBuffer {
-		watcher.add(event, c.timer, c.dispatchTimeoutBudget)
+		if !watcher.nonblockingAdd(event) {
+			c.blockedWatchers = append(c.blockedWatchers, watcher)
+		}
+	}
+
+	if len(c.blockedWatchers) > 0 {
+		// dispatchEvent is called very often, so arrange
+		// to reuse timers instead of constantly allocating.
+		startTime := time.Now()
+		timeout := c.dispatchTimeoutBudget.takeAvailable()
+		c.timer.Reset(timeout)
+
+		// Make sure every watcher will try to send event without blocking first,
+		// even if the timer has already expired.
+		for _, watcher := range c.blockedWatchers {
+			watcher.add(event, c.timer)
+		}
+		if !c.timer.Stop() {
```

Comment:
I just realized that this won't work.
The problem is that if one of the watcher times out, the next one won't get anything from the "timer.C" and may hang for a very long time.
It's possible to fix this problem by using context.WithTimeout() (that closes the underlying channel), that allocates timer on every call, so it's probably even worse.

So I don't know if we can fix this approach, so I'm afraid we probably should revert to what there was previously (unless you have better suggestions) - sorry for that.

Reply:
Achieved by a little more codes which makes the add() is similar to nonblocking().